// @bsv/sdk/dist/cjs/src/wallet/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validation = exports.WERR_INSUFFICIENT_FUNDS = exports.WERR_INVALID_PARAMETER = exports.WERR_REVIEW_ACTIONS = exports.WalletErrors = exports.WalletClient = exports.ProtoWallet = exports.CachedKeyDeriver = void 0;
__exportStar(require("./Wallet.interfaces.js"), exports);
__exportStar(require("./KeyDeriver.js"), exports);
var CachedKeyDeriver_js_1 = require("./CachedKeyDeriver.js");
Object.defineProperty(exports, "CachedKeyDeriver", { enumerable: true, get: function () { return __importDefault(CachedKeyDeriver_js_1).default; } });
var ProtoWallet_js_1 = require("./ProtoWallet.js");
Object.defineProperty(exports, "ProtoWallet", { enumerable: true, get: function () { return __importDefault(ProtoWallet_js_1).default; } });
var WalletClient_js_1 = require("./WalletClient.js");
Object.defineProperty(exports, "WalletClient", { enumerable: true, get: function () { return __importDefault(WalletClient_js_1).default; } });
// Is this an error? should it be 'walletErrors', the enum not the class?
var WalletError_js_1 = require("./WalletError.js");
Object.defineProperty(exports, "WalletErrors", { enumerable: true, get: function () { return __importDefault(WalletError_js_1).default; } });
var WERR_REVIEW_ACTIONS_js_1 = require("./WERR_REVIEW_ACTIONS.js");
Object.defineProperty(exports, "WERR_REVIEW_ACTIONS", { enumerable: true, get: function () { return __importDefault(WERR_REVIEW_ACTIONS_js_1).default; } });
var WERR_INVALID_PARAMETER_js_1 = require("./WERR_INVALID_PARAMETER.js");
Object.defineProperty(exports, "WERR_INVALID_PARAMETER", { enumerable: true, get: function () { return __importDefault(WERR_INVALID_PARAMETER_js_1).default; } });
var WERR_INSUFFICIENT_FUNDS_js_1 = require("./WERR_INSUFFICIENT_FUNDS.js");
Object.defineProperty(exports, "WERR_INSUFFICIENT_FUNDS", { enumerable: true, get: function () { return __importDefault(WERR_INSUFFICIENT_FUNDS_js_1).default; } });
__exportStar(require("./WalletError.js"), exports);
exports.Validation = __importStar(require("./validationHelpers.js"));
__exportStar(require("./substrates/index.js"), exports);
__exportStar(require("./WalletLoggerInterface.js"), exports);
